core/src/main/java/io/questdb/cairo/wal/SymbolMap.java

```diff
@@ -0,0 +1,213 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2025 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.wal;
+
+import io.questdb.std.AbstractOffsetCharSequenceHashSet;
+import io.questdb.std.Chars;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Numbers;
+import io.questdb.std.Unsafe;
+import io.questdb.std.str.DirectString;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Closeable;
+import java.util.Arrays;
+
+public class SymbolMap extends AbstractOffsetCharSequenceHashSet implements Closeable {
+    public static final int NO_ENTRY_VALUE = -1;
+    private final int noEntryValue;
+    private final DirectString sview = new DirectString();
+    private long address;
+    private long charsCapacity;
+    private int currentOffset = 0;
+    private int[] values;
```

Comment:
nit: we could go one step further and move `offsets` and `values` off-heap - that's to make the thing completely GC-free. Same story with hash codes - we could co-locate them with `offsets` like it's done in `OrderedMap` to avoid random memory accesses on key lookups.

Reply:
The changes look good, thanks.